src/devices/src/virtio/block/device.rs

```diff
@@ -713,8 +732,68 @@ pub(crate) mod tests {
 
             assert_eq!(vq.used.idx.get(), 1);
             assert_eq!(vq.used.ring[0].get().id, 0);
-            assert_eq!(vq.used.ring[0].get().len, vq.dtable[1].len.get());
+            // Added status descriptor length.
+            assert_eq!(vq.used.ring[0].get().len, vq.dtable[1].len.get() + 1);
             assert_eq!(mem.read_obj::<u32>(status_addr).unwrap(), VIRTIO_BLK_S_OK);
+        }
+
+        // Read with error.
+        {
+            vq.used.idx.set(0);
+            set_queue(&mut block, 0, vq.create_queue());
+
+            mem.write_obj::<u32>(VIRTIO_BLK_T_IN, request_type_addr)
+                .unwrap();
+            vq.dtable[1]
+                .flags
+                .set(VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE);
+
+            let size = block.disk.file.seek(SeekFrom::End(0)).unwrap();
+            block.disk.file.set_len(size / 2).unwrap();
+            mem.write_obj(10, GuestAddress(request_type_addr.0 + 8))
+                .unwrap();
+
+            invoke_handler_for_queue_event(&mut block);
+
+            assert_eq!(vq.used.idx.get(), 1);
+            assert_eq!(vq.used.ring[0].get().id, 0);
+
+            // Only status descriptor length.
+            assert_eq!(vq.used.ring[0].get().len, 1);
+            assert_eq!(
+                mem.read_obj::<u32>(status_addr).unwrap(),
+                VIRTIO_BLK_S_IOERR
+            );
+            assert_eq!(mem.read_obj::<u64>(data_addr).unwrap(), 123_456_789);
+        }
+
+        // Partial buffer error on read.
+        {
+            vq.used.idx.set(0);
+            set_queue(&mut block, 0, vq.create_queue());
+
+            mem.write_obj::<u32>(VIRTIO_BLK_T_IN, request_type_addr)
+                .unwrap();
+            vq.dtable[1]
+                .flags
+                .set(VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE);
+
+            let size = block.disk.file.seek(SeekFrom::End(0)).unwrap();
+            block.disk.file.set_len(size / 2).unwrap();
+            mem.write_obj(5, GuestAddress(request_type_addr.0 + 8))
```

Comment:
Can you add a comment here/somewhere about `request_type_addr.0 + 8` being the address where the sector number is stored? (It's not at all obvious the difference between this test case scenario and the previous one)?

Also, a comment why the length from the used ring entry = 1 would be also very useful (you would expect that for a partial buffer error on read, `completed` to be > 0, not = 0).

Reply:
We can also test the partial buffer error with `completed` > 0.
For example:

```rust
            vq.used.idx.set(0);
            set_queue(&mut block, 0, vq.create_queue());

            mem.write_obj::<u32>(VIRTIO_BLK_T_IN, request_type_addr)
                .unwrap();
            vq.dtable[1]
                .flags
                .set(VIRTQ_DESC_F_NEXT | VIRTQ_DESC_F_WRITE);
            vq.dtable[1].len.set(1024);

            let size = block.disk.file.seek(SeekFrom::End(0)).unwrap();
            block.disk.file.set_len(size / 2).unwrap();
            mem.write_obj(3, GuestAddress(request_type_addr.0 + 8))
                .unwrap();

            invoke_handler_for_queue_event(&mut block);

            assert_eq!(vq.used.idx.get(), 1);
            assert_eq!(vq.used.ring[0].get().id, 0);

            // File has 4 sectors and we try to read from sector 3, which means we will read
            // 512 bytes (instead of 1024).
            assert_eq!(vq.used.ring[0].get().len, 513);
            assert_eq!(
                mem.read_obj::<u32>(status_addr).unwrap(),
                VIRTIO_BLK_S_IOERR
            );
```

Reply:
@lauralt sounds good. Will you push this test in this PR or do you want me to do it?

Reply:
Sure, you can add it.